### PR TITLE
Issue/2414 create decorator for stable api

### LINKED ISSUE #2414: make abstraction of inmanta Python API elements
Methods that are part of the inmanta Python API are currently documented in `docs/reference/api.rst`. They are also tested against import loops in `tests/test_import_entry_points.py`. Keeping track of this while developing is tedious now and might become difficult as the size of the API scales. Making abstraction of which elements are part of the API in some structured way could centralize this information. This could provide multiple advantages:
- API documentation and entry point import tests could be automated. This would guarantee they are always up to date and can no longer be forgotten.
- Checks could be built to protect/warn against changes to the API interface. Currently, a refactor that changes the interface could be merged in without any test failures or warnings. This could cause tools/modules/extensions that depend on this interface to fail.

A decentralized collection approach (using a decorator for example) could be even better but that might not be feasible.

I understand this description is still somewhat vague. This is partially a design issue. The main goal in my opinion would be to protect against changes to the interface.

# ideas

1. add a decorator to indicate different types of api 
2. annotate the current documented api's (`docs/reference/api.rst`  and `docs/reference/json.rst`)

# Stage 2

create ticket to 
2. use to generate import testcase
3. use to generate docs  (see inmanta-sphinx)

### PR DESCRIPTION
**Pull request by arnaudsjs**
# Description

I added the `stable_api` decorator on the classes and functions mentioned on the documentation pages that describe the stable API.

**Questions:**

* I didn't annotate the methods in the `methods.py` and the `methods_v2.py`, because these methods are already tracked via the `@method` decorator. Should I also add the `public_api` decorator to them? 
* What about variables which are not part of a class, such as `inmanta.module.INSTALL_OPTS`, `inmanta.ast.type.TYPES`, `inmanta.data.TBaseDocument`, `inmanta.data.model.ResourceIdStr` and `inmanta.data.model.ResourceVersionIdStr`. How are we going to keep track of them? Wrap them in an method invocation?

closes #2414 

# Self Check:

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/decorator.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+from typing import Callable, Type, Union
+
+
+def stable_api(func: Union[Callable, Type]) -> Union[Callable, Type]:
```
Comment (wouterdb) [inline src/inmanta/decorator.py:21]: should n't this be 

```suggestion
def stable_api(func: T) -> T:
```

with T bounded on `Union[Callable, Type]`

**Review by wouterdb — COMMENTED** [review]
1. I wouldn't do the methods twice, as they clearly are API
2. either package a copy of these values into object or add a comment?

**Comment by arnaudsjs** [comment]
> either package a copy of these values into a public object or add a comment to them?

I decided to add a comment.

**Review by wouterdb — APPROVED** [review]

**Review by sanderr — APPROVED** [review]
It doesn't declare which fields/methods of a class are part of the API and which are considered internal but I guess for this stage this suffices. If the annotation is there, we could say it's up to the developer to have a look at the docs and determine which parts they are free to change.

**Comment by sanderr** [comment]
I agree with Wouter on both points. Though we should find a better way to deal with plain variables in the future.

**Pull request closed without merging**